{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "random_sampling.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "authorship_tag": "ABX9TyO/x+pBOu+KeenIJUld0Lvd",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/alexchen1999/covid-19-sample-strain-classification/blob/main/random_sampling.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install Bio\n",
        "!pip install bcbio-gff\n",
        "!pip install ncbi-acc-download"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iDtyId95jj2_",
        "outputId": "fed04a67-dc66-45db-b2e7-db7035170528"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting Bio\n",
            "  Downloading bio-1.3.9-py3-none-any.whl (270 kB)\n",
            "\u001b[K     |████████████████████████████████| 270 kB 5.2 MB/s \n",
            "\u001b[?25hCollecting mygene\n",
            "  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.7/dist-packages (from Bio) (4.64.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from Bio) (2.23.0)\n",
            "Collecting biopython>=1.79\n",
            "  Downloading biopython-1.79-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (2.3 MB)\n",
            "\u001b[K     |████████████████████████████████| 2.3 MB 39.2 MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython>=1.79->Bio) (1.21.6)\n",
            "Collecting biothings-client>=0.2.6\n",
            "  Downloading biothings_client-0.2.6-py2.py3-none-any.whl (37 kB)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (3.0.4)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2.10)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (1.24.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2022.6.15)\n",
            "Installing collected packages: biothings-client, mygene, biopython, Bio\n",
            "Successfully installed Bio-1.3.9 biopython-1.79 biothings-client-0.2.6 mygene-3.2.2\n",
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting bcbio-gff\n",
            "  Downloading bcbio-gff-0.6.9.tar.gz (44 kB)\n",
            "\u001b[K     |████████████████████████████████| 44 kB 2.0 MB/s \n",
            "\u001b[?25hRequirement already satisfied: six in /usr/local/lib/python3.7/dist-packages (from bcbio-gff) (1.15.0)\n",
            "Requirement already satisfied: biopython in /usr/local/lib/python3.7/dist-packages (from bcbio-gff) (1.79)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython->bcbio-gff) (1.21.6)\n",
            "Building wheels for collected packages: bcbio-gff\n",
            "  Building wheel for bcbio-gff (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for bcbio-gff: filename=bcbio_gff-0.6.9-py3-none-any.whl size=15954 sha256=93867b13d582c02e5fefc647d369cb434c76e2e042d0adbd0162ccf406c309b5\n",
            "  Stored in directory: /root/.cache/pip/wheels/fb/fc/e0/2e7658046d32b794fdfd1e0ec266dc8bae21c1811b18951b6d\n",
            "Successfully built bcbio-gff\n",
            "Installing collected packages: bcbio-gff\n",
            "Successfully installed bcbio-gff-0.6.9\n",
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting ncbi-acc-download\n",
            "  Downloading ncbi_acc_download-0.2.8-py2.py3-none-any.whl (17 kB)\n",
            "Requirement already satisfied: requests>=2.4.3 in /usr/local/lib/python3.7/dist-packages (from ncbi-acc-download) (2.23.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests>=2.4.3->ncbi-acc-download) (2022.6.15)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests>=2.4.3->ncbi-acc-download) (1.24.3)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests>=2.4.3->ncbi-acc-download) (3.0.4)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests>=2.4.3->ncbi-acc-download) (2.10)\n",
            "Installing collected packages: ncbi-acc-download\n",
            "Successfully installed ncbi-acc-download-0.2.8\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {
        "id": "pqUtK0J3aL7O"
      },
      "outputs": [],
      "source": [
        "import json\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import subprocess\n",
        "\n",
        "from BCBio import GFF\n",
        "from Bio import Align, SeqIO\n",
        "from numpy import array\n",
        "from numpy import argmax\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "from sklearn.preprocessing import OneHotEncoder\n",
        "from sklearn.utils import shuffle"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "covid_strains = ['B.1.1.7_sequences.csv', 'P.1_sequences.csv', 'B.1.617.2_sequences.csv', 'BA.1.1_sequences.csv']"
      ],
      "metadata": {
        "id": "bHZT6tibTThF"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "alpha = pd.read_csv(covid_strains[0])\n",
        "gamma = pd.read_csv(covid_strains[1])\n",
        "delta = pd.read_csv(covid_strains[2])\n",
        "omicron = pd.read_csv(covid_strains[3])"
      ],
      "metadata": {
        "id": "r_7tx1WBT4Qm"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "shortest = min([alpha.shape[0], gamma.shape[0], delta.shape[0], omicron.shape[0]])\n",
        "print(shortest)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "kejANgblZDUj",
        "outputId": "f52a3ea1-7294-452b-8ce3-59776beccc88"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3735\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "alpha_sample = alpha.sample(n=shortest)\n",
        "print(alpha_sample.shape[0])\n",
        "gamma_sample = gamma.sample(n=shortest)\n",
        "print(gamma_sample.shape[0])\n",
        "delta_sample = delta.sample(n=shortest)\n",
        "print(delta_sample.shape[0])\n",
        "omicron_sample = omicron.sample(n=shortest)\n",
        "print(omicron_sample.shape[0])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "UvnW3PaZcA1d",
        "outputId": "867cf52b-8e03-4492-ee54-106c413d6746"
      },
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3735\n",
            "3735\n",
            "3735\n",
            "3735\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "print(alpha_sample['Country'].unique())\n",
        "print(gamma_sample['Country'].unique())\n",
        "print(delta_sample['Country'].unique())\n",
        "print(omicron_sample['Country'].unique())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7QM1y0xSe14h",
        "outputId": "0ea72ae0-c571-4471-8899-2472fa8c3909"
      },
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "['USA' 'Iraq' 'Japan' 'Pakistan' 'Finland' 'Nigeria' 'Austria' 'Djibouti'\n",
            " 'Spain' 'India' 'Egypt' 'Philippines' 'West Bank' 'Bangladesh' 'Denmark'\n",
            " 'Mexico' 'Italy']\n",
            "['USA' 'Dominican Republic' 'Brazil' 'Paraguay' 'Mexico' 'Taiwan' 'Chile'\n",
            " 'Italy' 'Peru']\n",
            "['USA' 'Bahrain' 'Bangladesh' 'Egypt' 'Japan' 'Mongolia' 'India' 'Myanmar'\n",
            " 'Pakistan' 'Chile' 'Uzbekistan' 'Gabon' 'Russia' 'Denmark' 'Jamaica'\n",
            " 'West Bank' 'China']\n",
            "['USA' 'Bahrain' 'West Bank' 'France' 'Japan']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# From SARS-Cov-2 GFF file, find genomic location of Spike Glycoprotein\n",
        "in_file = \"GCF_009858895.2_ASM985889v3_genomic.gff\"\n",
        "in_handle = open(in_file)\n",
        "features = []\n",
        "for rec in GFF.parse(in_handle):\n",
        "    features = rec.features\n",
        "in_handle.close()\n",
        "\n",
        "start = -1\n",
        "end = -1\n",
        "\n",
        "# spike glycoprotein locus tag GU280_gp02\n",
        "for i in range(len(features)):\n",
        "    if \"GU280_gp02\" in features[i].id:\n",
        "      print(features[i])\n",
        "      print(features[i].location)\n",
        "      print(int(features[i].location._start))\n",
        "      print(int(features[i].location._end))\n",
        "      start = int(features[i].location._start)\n",
        "      end = int(features[i].location._end)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "l3Cv8cyG-t-k",
        "outputId": "5373910f-7f76-498b-8c5a-ee6d71187ae5"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "type: gene\n",
            "location: [21562:25384](+)\n",
            "id: gene-GU280_gp02\n",
            "qualifiers:\n",
            "    Key: Dbxref, Value: ['GeneID:43740568']\n",
            "    Key: ID, Value: ['gene-GU280_gp02']\n",
            "    Key: Name, Value: ['S']\n",
            "    Key: gbkey, Value: ['Gene']\n",
            "    Key: gene, Value: ['S']\n",
            "    Key: gene_biotype, Value: ['protein_coding']\n",
            "    Key: gene_synonym, Value: ['spike glycoprotein']\n",
            "    Key: locus_tag, Value: ['GU280_gp02']\n",
            "    Key: source, Value: ['RefSeq']\n",
            "\n",
            "[21562:25384](+)\n",
            "21562\n",
            "25384\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Find spike glycoprotein sequence from reference genome\n",
        "record = SeqIO.read(\"NC_045512.2.fa\", \"fasta\")\n",
        "\n",
        "# Take from 21562 to 25384\n",
        "spike_seq = str(record._seq)[start:end]\n",
        "print(spike_seq)\n",
        "print(len(spike_seq))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_AX75FXDU1-q",
        "outputId": "3b131245-8196-420e-d348-5e532e7d9e3b"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ATGTTTGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTGTTAATCTTACAACCAGAACTCAATTACCCCCTGCATACACTAATTCTTTCACACGTGGTGTTTATTACCCTGACAAAGTTTTCAGATCCTCAGTTTTACATTCAACTCAGGACTTGTTCTTACCTTTCTTTTCCAATGTTACTTGGTTCCATGCTATACATGTCTCTGGGACCAATGGTACTAAGAGGTTTGATAACCCTGTCCTACCATTTAATGATGGTGTTTATTTTGCTTCCACTGAGAAGTCTAACATAATAAGAGGCTGGATTTTTGGTACTACTTTAGATTCGAAGACCCAGTCCCTACTTATTGTTAATAACGCTACTAATGTTGTTATTAAAGTCTGTGAATTTCAATTTTGTAATGATCCATTTTTGGGTGTTTATTACCACAAAAACAACAAAAGTTGGATGGAAAGTGAGTTCAGAGTTTATTCTAGTGCGAATAATTGCACTTTTGAATATGTCTCTCAGCCTTTTCTTATGGACCTTGAAGGAAAACAGGGTAATTTCAAAAATCTTAGGGAATTTGTGTTTAAGAATATTGATGGTTATTTTAAAATATATTCTAAGCACACGCCTATTAATTTAGTGCGTGATCTCCCTCAGGGTTTTTCGGCTTTAGAACCATTGGTAGATTTGCCAATAGGTATTAACATCACTAGGTTTCAAACTTTACTTGCTTTACATAGAAGTTATTTGACTCCTGGTGATTCTTCTTCAGGTTGGACAGCTGGTGCTGCAGCTTATTATGTGGGTTATCTTCAACCTAGGACTTTTCTATTAAAATATAATGAAAATGGAACCATTACAGATGCTGTAGACTGTGCACTTGACCCTCTCTCAGAAACAAAGTGTACGTTGAAATCCTTCACTGTAGAAAAAGGAATCTATCAAACTTCTAACTTTAGAGTCCAACCAACAGAATCTATTGTTAGATTTCCTAATATTACAAACTTGTGCCCTTTTGGTGAAGTTTTTAACGCCACCAGATTTGCATCTGTTTATGCTTGGAACAGGAAGAGAATCAGCAACTGTGTTGCTGATTATTCTGTCCTATATAATTCCGCATCATTTTCCACTTTTAAGTGTTATGGAGTGTCTCCTACTAAATTAAATGATCTCTGCTTTACTAATGTCTATGCAGATTCATTTGTAATTAGAGGTGATGAAGTCAGACAAATCGCTCCAGGGCAAACTGGAAAGATTGCTGATTATAATTATAAATTACCAGATGATTTTACAGGCTGCGTTATAGCTTGGAATTCTAACAATCTTGATTCTAAGGTTGGTGGTAATTATAATTACCTGTATAGATTGTTTAGGAAGTCTAATCTCAAACCTTTTGAGAGAGATATTTCAACTGAAATCTATCAGGCCGGTAGCACACCTTGTAATGGTGTTGAAGGTTTTAATTGTTACTTTCCTTTACAATCATATGGTTTCCAACCCACTAATGGTGTTGGTTACCAACCATACAGAGTAGTAGTACTTTCTTTTGAACTTCTACATGCACCAGCAACTGTTTGTGGACCTAAAAAGTCTACTAATTTGGTTAAAAACAAATGTGTCAATTTCAACTTCAATGGTTTAACAGGCACAGGTGTTCTTACTGAGTCTAACAAAAAGTTTCTGCCTTTCCAACAATTTGGCAGAGACATTGCTGACACTACTGATGCTGTCCGTGATCCACAGACACTTGAGATTCTTGACATTACACCATGTTCTTTTGGTGGTGTCAGTGTTATAACACCAGGAACAAATACTTCTAACCAGGTTGCTGTTCTTTATCAGGATGTTAACTGCACAGAAGTCCCTGTTGCTATTCATGCAGATCAACTTACTCCTACTTGGCGTGTTTATTCTACAGGTTCTAATGTTTTTCAAACACGTGCAGGCTGTTTAATAGGGGCTGAACATGTCAACAACTCATATGAGTGTGACATACCCATTGGTGCAGGTATATGCGCTAGTTATCAGACTCAGACTAATTCTCCTCGGCGGGCACGTAGTGTAGCTAGTCAATCCATCATTGCCTACACTATGTCACTTGGTGCAGAAAATTCAGTTGCTTACTCTAATAACTCTATTGCCATACCCACAAATTTTACTATTAGTGTTACCACAGAAATTCTACCAGTGTCTATGACCAAGACATCAGTAGATTGTACAATGTACATTTGTGGTGATTCAACTGAATGCAGCAATCTTTTGTTGCAATATGGCAGTTTTTGTACACAATTAAACCGTGCTTTAACTGGAATAGCTGTTGAACAAGACAAAAACACCCAAGAAGTTTTTGCACAAGTCAAACAAATTTACAAAACACCACCAATTAAAGATTTTGGTGGTTTTAATTTTTCACAAATATTACCAGATCCATCAAAACCAAGCAAGAGGTCATTTATTGAAGATCTACTTTTCAACAAAGTGACACTTGCAGATGCTGGCTTCATCAAACAATATGGTGATTGCCTTGGTGATATTGCTGCTAGAGACCTCATTTGTGCACAAAAGTTTAACGGCCTTACTGTTTTGCCACCTTTGCTCACAGATGAAATGATTGCTCAATACACTTCTGCACTGTTAGCGGGTACAATCACTTCTGGTTGGACCTTTGGTGCAGGTGCTGCATTACAAATACCATTTGCTATGCAAATGGCTTATAGGTTTAATGGTATTGGAGTTACACAGAATGTTCTCTATGAGAACCAAAAATTGATTGCCAACCAATTTAATAGTGCTATTGGCAAAATTCAAGACTCACTTTCTTCCACAGCAAGTGCACTTGGAAAACTTCAAGATGTGGTCAACCAAAATGCACAAGCTTTAAACACGCTTGTTAAACAACTTAGCTCCAATTTTGGTGCAATTTCAAGTGTTTTAAATGATATCCTTTCACGTCTTGACAAAGTTGAGGCTGAAGTGCAAATTGATAGGTTGATCACAGGCAGACTTCAAAGTTTGCAGACATATGTGACTCAACAATTAATTAGAGCTGCAGAAATCAGAGCTTCTGCTAATCTTGCTGCTACTAAAATGTCAGAGTGTGTACTTGGACAATCAAAAAGAGTTGATTTTTGTGGAAAGGGCTATCATCTTATGTCCTTCCCTCAGTCAGCACCTCATGGTGTAGTCTTCTTGCATGTGACTTATGTCCCTGCACAAGAAAAGAACTTCACAACTGCTCCTGCCATTTGTCATGATGGAAAAGCACACTTTCCTCGTGAAGGTGTCTTTGTTTCAAATGGCACACACTGGTTTGTAACACAAAGGAATTTTTATGAACCACAAATCATTACTACAGACAACACATTTGTGTCTGGTAACTGTGATGTTGTAATAGGAATTGTCAACAACACAGTTTATGATCCTTTGCAACCTGAATTAGACTCATTCAAGGAGGAGTTAGATAAATATTTTAAGAATCATACATCACCAGATGTTGATTTAGGTGACATCTCTGGCATTAATGCTTCAGTTGTAAACATTCAAAAAGAAATTGACCGCCTCAATGAGGTTGCCAAGAATTTAAATGAATCTCTCATCGATCTCCAAGAACTTGGAAAGTATGAGCAGTATATAAAATGGCCATGGTACATTTGGCTAGGTTTTATAGCTGGCTTGATTGCCATAGTAATGGTGACAATTATGCTTTGCTGTATGACCAGTTGCTGTAGTTGTCTCAAGGGCTGTTGTTCTTGTGGATCCTGCTGCAAATTTGATGAAGACGACTCTGAGCCAGTGCTCAAAGGAGTCAAATTACATTACACATAA\n",
            "3822\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "alpha_sample_accessions = []\n",
        "for i in range(shortest):\n",
        "  alpha_sample_accessions.append(alpha_sample.iloc[i]['Accession'])"
      ],
      "metadata": {
        "id": "8cBPDZI8gsaK"
      },
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!ncbi-acc-download --help"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KB02eVCcjmoy",
        "outputId": "02cc7146-7c92-4ef3-c814-eed74efd9d06"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "usage: ncbi-acc-download [-h] [-m {nucleotide,protein}] [--api-key API_KEY]\n",
            "                         [-e {none,loads,all,correct}]\n",
            "                         [-F {fasta,genbank,featuretable,gff3}] [-o OUT]\n",
            "                         [-p PREFIX] [-g RANGE] [-r] [--url] [-v]\n",
            "                         NCBI-accession [NCBI-accession ...]\n",
            "\n",
            "positional arguments:\n",
            "  NCBI-accession\n",
            "\n",
            "optional arguments:\n",
            "  -h, --help            show this help message and exit\n",
            "  -m {nucleotide,protein}, --molecule {nucleotide,protein}\n",
            "                        Molecule type to download. Default: nucleotide\n",
            "  --api-key API_KEY     Specify USER NCBI API key. More info at\n",
            "                        https://www.ncbi.nlm.nih.gov/books/NBK25497/\n",
            "  -e {none,loads,all,correct}, --extended-validation {none,loads,all,correct}\n",
            "                        Perform extended validation. Possible options are\n",
            "                        'none' to skip validation, 'loads' to check if the\n",
            "                        sequence file loads in Biopython, or 'all' to run all\n",
            "                        checks. Default: none\n",
            "  -F {fasta,genbank,featuretable,gff3}, --format {fasta,genbank,featuretable,gff3}\n",
            "                        File format to download nucleotide sequences in.\n",
            "                        Default: genbank\n",
            "  -o OUT, --out OUT     Single filename to use for the combined output.\n",
            "  -p PREFIX, --prefix PREFIX\n",
            "                        Filename prefix to use for output files instead of\n",
            "                        using the NCBI ID.\n",
            "  -g RANGE, --range RANGE\n",
            "                        region to subset accession. only for single accession\n",
            "  -r, --recursive       Recursively get all entries of a WGS entry.\n",
            "  --url                 Instead of downloading the sequences, just print the\n",
            "                        URLs to stdout.\n",
            "  -v, --verbose         Print a progress indicator.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Test NCBI accession download\n",
        "test = alpha_sample_accessions[0]\n",
        "print(test)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "uAUkAs_7mGKZ",
        "outputId": "6104c47d-b216-4ae1-f610-ae3e4e8196a8"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "MZ164127.1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!ncbi-acc-download --format fasta $test"
      ],
      "metadata": {
        "id": "M5f4wYb_kS2l"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "for i in range(3):\n",
        "  subprocess.run([\"ncbi-acc-download\", \"--format\", \"fasta\", alpha_sample_accessions[i]])"
      ],
      "metadata": {
        "id": "WM5QZHm9oOoy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# SeqIO\n",
        "record = SeqIO.read(\"MZ039190.1.fa\", \"fasta\")\n",
        "seq = str(record._seq)[start:end]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 353
        },
        "id": "km0c1zm5slUu",
        "outputId": "419f06d9-272d-441e-b3f9-1a7175228914"
      },
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-13-b9110862e65f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# SeqIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"MZ039190.1.fa\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mseq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecord\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_seq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/Bio/SeqIO/__init__.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    650\u001b[0m     \u001b[0mto\u001b[0m \u001b[0mread\u001b[0m \u001b[0mmultiple\u001b[0m \u001b[0mrecords\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    651\u001b[0m     \"\"\"\n\u001b[0;32m--> 652\u001b[0;31m     \u001b[0miterator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malphabet\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    653\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    654\u001b[0m         \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0miterator\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/Bio/SeqIO/__init__.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    603\u001b[0m     \u001b[0miterator_generator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_FormatToIterator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    604\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0miterator_generator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 605\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0miterator_generator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    606\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mformat\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mAlignIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_FormatToIterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    607\u001b[0m         \u001b[0;31m# Use Bio.AlignIO to read in the alignments\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/Bio/SeqIO/FastaIO.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, source, alphabet, title2ids)\u001b[0m\n\u001b[1;32m    181\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"The alphabet argument is no longer supported\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    182\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtitle2ids\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtitle2ids\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 183\u001b[0;31m         \u001b[0msuper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"t\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfmt\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"Fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    184\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/Bio/SeqIO/Interfaces.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, source, alphabet, mode, fmt)\u001b[0m\n\u001b[1;32m     46\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"The alphabet argument is no longer supported\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 48\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     49\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshould_close_stream\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     50\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# not a path, assume we received a stream\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'MZ039190.1.fa'"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Pairwise alignment\n",
        "aligner = Align.PairwiseAligner()\n",
        "aligner.mode = 'global'\n",
        "alignments = aligner.align(seq, spike_seq)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 204
        },
        "id": "pBYp4YciMDmI",
        "outputId": "1ab54ad7-34c5-4c93-eda3-2d4ec457e554"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-14-f56c135df7bd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0maligner\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mAlign\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPairwiseAligner\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0maligner\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'global'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0malignments\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maligner\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malign\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mspike_seq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m: name 'seq' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Alpha sample accessions\n",
        "# Take from 21562 to 25384 from samples, should be a good enough approximation\n",
        "# Credit to https://2-bitbio.com/2018/06/one-hot-encode-dna-sequence-using.html\n",
        "\n",
        "alpha_spike_seq_one_hot = pd.DataFrame(columns=[\"Accession\", \"Seq\"])\n",
        "alpha_spike_seq_df = pd.DataFrame(columns = ['Accession', 'Seq'])\n",
        "for i in range(3):\n",
        "  accession = alpha_sample_accessions[i]\n",
        "  subprocess.run([\"ncbi-acc-download\", \"--format\", \"fasta\", accession])\n",
        "  record = SeqIO.read(accession + \".fa\", \"fasta\")\n",
        "  spike_seq = str(record._seq)[start:end]\n",
        "\n",
        "  # One-hot encoding\n",
        "  seq_array = array(list(spike_seq))\n",
        "  label_encoder = LabelEncoder()\n",
        "  integer_encoded_seq = label_encoder.fit_transform(seq_array)\n",
        "    \n",
        "  # one hot the sequence\n",
        "  onehot_encoder = OneHotEncoder(sparse=False)\n",
        "  # reshape because that's what OneHotEncoder likes\n",
        "  integer_encoded_seq = integer_encoded_seq.reshape(len(integer_encoded_seq), 1)\n",
        "  onehot_encoded_seq = onehot_encoder.fit_transform(integer_encoded_seq)\n",
        "\n",
        "  # convert the sequence from ndarray to json string to simplify conversion\n",
        "  onehot_encoded_seq = json.dumps(onehot_encoded_seq.tolist())\n",
        "\n",
        "  alpha_spike_seq_df.loc[len(alpha_spike_seq_df.index)] = [accession, onehot_encoded_seq]\n",
        "  subprocess.run(['rm', accession + \".fa\"])"
      ],
      "metadata": {
        "id": "cTaIdMvBPCFG"
      },
      "execution_count": 16,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# All\n",
        "dfs = [alpha_sample, delta_sample, gamma_sample, omicron_sample]\n",
        "labels = ['Alpha', 'Delta', 'Gamma', 'Omicron']\n",
        "spike_seq_df = pd.DataFrame(columns = ['Accession', 'Seq', 'Label'])\n",
        "num_samples = 1000\n",
        "for i in range(num_samples):\n",
        "  for j in range(4):\n",
        "    accession = dfs[j].iloc[i]['Accession']\n",
        "\n",
        "    subprocess.run([\"ncbi-acc-download\", \"--format\", \"fasta\", accession])\n",
        "    record = SeqIO.read(accession + \".fa\", \"fasta\")\n",
        "    spike_seq = str(record._seq)[start:end]\n",
        "\n",
        "    # One-hot encoding\n",
        "    seq_array = array(list(spike_seq))\n",
        "    label_encoder = LabelEncoder()\n",
        "    integer_encoded_seq = label_encoder.fit_transform(seq_array)\n",
        "    \n",
        "    #one hot the sequence\n",
        "    onehot_encoder = OneHotEncoder(sparse=False)\n",
        "    #reshape because that's what OneHotEncoder likes\n",
        "    integer_encoded_seq = integer_encoded_seq.reshape(len(integer_encoded_seq), 1)\n",
        "    onehot_encoded_seq = onehot_encoder.fit_transform(integer_encoded_seq)\n",
        "\n",
        "    # convert the sequence from ndarray to json string to simplify conversion\n",
        "    onehot_encoded_seq = json.dumps(onehot_encoded_seq.tolist())\n",
        "\n",
        "    spike_seq_df.loc[len(spike_seq_df.index)] = [accession, onehot_encoded_seq, labels[j]]\n",
        "    subprocess.run(['rm', accession + \".fa\"])\n",
        "  print((i+1) / num_samples * 100, \"% complete\")\n",
        "\n",
        "df = shuffle(spike_seq_df)\n",
        "df.to_csv('spike_seq_one_hots.csv')"
      ],
      "metadata": {
        "id": "qc8DPJB71pPw",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "2ae360be-7dbf-4970-c382-bb75393d0734"
      },
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "0.1 % complete\n",
            "0.2 % complete\n",
            "0.3 % complete\n",
            "0.4 % complete\n",
            "0.5 % complete\n",
            "0.6 % complete\n",
            "0.7000000000000001 % complete\n",
            "0.8 % complete\n",
            "0.8999999999999999 % complete\n",
            "1.0 % complete\n",
            "1.0999999999999999 % complete\n",
            "1.2 % complete\n",
            "1.3 % complete\n",
            "1.4000000000000001 % complete\n",
            "1.5 % complete\n",
            "1.6 % complete\n",
            "1.7000000000000002 % complete\n",
            "1.7999999999999998 % complete\n",
            "1.9 % complete\n",
            "2.0 % complete\n",
            "2.1 % complete\n",
            "2.1999999999999997 % complete\n",
            "2.3 % complete\n",
            "2.4 % complete\n",
            "2.5 % complete\n",
            "2.6 % complete\n",
            "2.7 % complete\n",
            "2.8000000000000003 % complete\n",
            "2.9000000000000004 % complete\n",
            "3.0 % complete\n",
            "3.1 % complete\n",
            "3.2 % complete\n",
            "3.3000000000000003 % complete\n",
            "3.4000000000000004 % complete\n",
            "3.5000000000000004 % complete\n",
            "3.5999999999999996 % complete\n",
            "3.6999999999999997 % complete\n",
            "3.8 % complete\n",
            "3.9 % complete\n",
            "4.0 % complete\n",
            "4.1000000000000005 % complete\n",
            "4.2 % complete\n",
            "4.3 % complete\n",
            "4.3999999999999995 % complete\n",
            "4.5 % complete\n",
            "4.6 % complete\n",
            "4.7 % complete\n",
            "4.8 % complete\n",
            "4.9 % complete\n",
            "5.0 % complete\n",
            "5.1 % complete\n",
            "5.2 % complete\n",
            "5.3 % complete\n",
            "5.4 % complete\n",
            "5.5 % complete\n",
            "5.6000000000000005 % complete\n",
            "5.7 % complete\n",
            "5.800000000000001 % complete\n",
            "5.8999999999999995 % complete\n",
            "6.0 % complete\n",
            "6.1 % complete\n",
            "6.2 % complete\n",
            "6.3 % complete\n",
            "6.4 % complete\n",
            "6.5 % complete\n",
            "6.6000000000000005 % complete\n",
            "6.7 % complete\n",
            "6.800000000000001 % complete\n",
            "6.9 % complete\n",
            "7.000000000000001 % complete\n",
            "7.1 % complete\n",
            "7.199999999999999 % complete\n",
            "7.3 % complete\n",
            "7.3999999999999995 % complete\n",
            "7.5 % complete\n",
            "7.6 % complete\n",
            "7.7 % complete\n",
            "7.8 % complete\n",
            "7.9 % complete\n",
            "8.0 % complete\n",
            "8.1 % complete\n",
            "8.200000000000001 % complete\n",
            "8.3 % complete\n",
            "8.4 % complete\n",
            "8.5 % complete\n",
            "8.6 % complete\n",
            "8.7 % complete\n",
            "8.799999999999999 % complete\n",
            "8.9 % complete\n",
            "9.0 % complete\n",
            "9.1 % complete\n",
            "9.2 % complete\n",
            "9.3 % complete\n",
            "9.4 % complete\n",
            "9.5 % complete\n",
            "9.6 % complete\n",
            "9.700000000000001 % complete\n",
            "9.8 % complete\n",
            "9.9 % complete\n",
            "10.0 % complete\n",
            "10.100000000000001 % complete\n",
            "10.2 % complete\n",
            "10.299999999999999 % complete\n",
            "10.4 % complete\n",
            "10.5 % complete\n",
            "10.6 % complete\n",
            "10.7 % complete\n",
            "10.8 % complete\n",
            "10.9 % complete\n",
            "11.0 % complete\n",
            "11.1 % complete\n",
            "11.200000000000001 % complete\n",
            "11.3 % complete\n",
            "11.4 % complete\n",
            "11.5 % complete\n",
            "11.600000000000001 % complete\n",
            "11.700000000000001 % complete\n",
            "11.799999999999999 % complete\n",
            "11.899999999999999 % complete\n",
            "12.0 % complete\n",
            "12.1 % complete\n",
            "12.2 % complete\n",
            "12.3 % complete\n",
            "12.4 % complete\n",
            "12.5 % complete\n",
            "12.6 % complete\n",
            "12.7 % complete\n",
            "12.8 % complete\n",
            "12.9 % complete\n",
            "13.0 % complete\n",
            "13.100000000000001 % complete\n",
            "13.200000000000001 % complete\n",
            "13.3 % complete\n",
            "13.4 % complete\n",
            "13.5 % complete\n",
            "13.600000000000001 % complete\n",
            "13.700000000000001 % complete\n",
            "13.8 % complete\n",
            "13.900000000000002 % complete\n",
            "14.000000000000002 % complete\n",
            "14.099999999999998 % complete\n",
            "14.2 % complete\n",
            "14.299999999999999 % complete\n",
            "14.399999999999999 % complete\n",
            "14.499999999999998 % complete\n",
            "14.6 % complete\n",
            "14.7 % complete\n",
            "14.799999999999999 % complete\n",
            "14.899999999999999 % complete\n",
            "15.0 % complete\n",
            "15.1 % complete\n",
            "15.2 % complete\n",
            "15.299999999999999 % complete\n",
            "15.4 % complete\n",
            "15.5 % complete\n",
            "15.6 % complete\n",
            "15.7 % complete\n",
            "15.8 % complete\n",
            "15.9 % complete\n",
            "16.0 % complete\n",
            "16.1 % complete\n",
            "16.2 % complete\n",
            "16.3 % complete\n",
            "16.400000000000002 % complete\n",
            "16.5 % complete\n",
            "16.6 % complete\n",
            "16.7 % complete\n",
            "16.8 % complete\n",
            "16.900000000000002 % complete\n",
            "17.0 % complete\n",
            "17.1 % complete\n",
            "17.2 % complete\n",
            "17.299999999999997 % complete\n",
            "17.4 % complete\n",
            "17.5 % complete\n",
            "17.599999999999998 % complete\n",
            "17.7 % complete\n",
            "17.8 % complete\n",
            "17.9 % complete\n",
            "18.0 % complete\n",
            "18.099999999999998 % complete\n",
            "18.2 % complete\n",
            "18.3 % complete\n",
            "18.4 % complete\n",
            "18.5 % complete\n",
            "18.6 % complete\n",
            "18.7 % complete\n",
            "18.8 % complete\n",
            "18.9 % complete\n",
            "19.0 % complete\n",
            "19.1 % complete\n",
            "19.2 % complete\n",
            "19.3 % complete\n",
            "19.400000000000002 % complete\n",
            "19.5 % complete\n",
            "19.6 % complete\n",
            "19.7 % complete\n",
            "19.8 % complete\n",
            "19.900000000000002 % complete\n",
            "20.0 % complete\n",
            "20.1 % complete\n",
            "20.200000000000003 % complete\n",
            "20.3 % complete\n",
            "20.4 % complete\n",
            "20.5 % complete\n",
            "20.599999999999998 % complete\n",
            "20.7 % complete\n",
            "20.8 % complete\n",
            "20.9 % complete\n",
            "21.0 % complete\n",
            "21.099999999999998 % complete\n",
            "21.2 % complete\n",
            "21.3 % complete\n",
            "21.4 % complete\n",
            "21.5 % complete\n",
            "21.6 % complete\n",
            "21.7 % complete\n",
            "21.8 % complete\n",
            "21.9 % complete\n",
            "22.0 % complete\n",
            "22.1 % complete\n",
            "22.2 % complete\n",
            "22.3 % complete\n",
            "22.400000000000002 % complete\n",
            "22.5 % complete\n",
            "22.6 % complete\n",
            "22.7 % complete\n",
            "22.8 % complete\n",
            "22.900000000000002 % complete\n",
            "23.0 % complete\n",
            "23.1 % complete\n",
            "23.200000000000003 % complete\n",
            "23.3 % complete\n",
            "23.400000000000002 % complete\n",
            "23.5 % complete\n",
            "23.599999999999998 % complete\n",
            "23.7 % complete\n",
            "23.799999999999997 % complete\n",
            "23.9 % complete\n",
            "24.0 % complete\n",
            "24.099999999999998 % complete\n",
            "24.2 % complete\n",
            "24.3 % complete\n",
            "24.4 % complete\n",
            "24.5 % complete\n",
            "24.6 % complete\n",
            "24.7 % complete\n",
            "24.8 % complete\n",
            "24.9 % complete\n",
            "25.0 % complete\n",
            "25.1 % complete\n",
            "25.2 % complete\n",
            "25.3 % complete\n",
            "25.4 % complete\n",
            "25.5 % complete\n",
            "25.6 % complete\n",
            "25.7 % complete\n",
            "25.8 % complete\n",
            "25.900000000000002 % complete\n",
            "26.0 % complete\n",
            "26.1 % complete\n",
            "26.200000000000003 % complete\n",
            "26.3 % complete\n",
            "26.400000000000002 % complete\n",
            "26.5 % complete\n",
            "26.6 % complete\n",
            "26.700000000000003 % complete\n",
            "26.8 % complete\n",
            "26.900000000000002 % complete\n",
            "27.0 % complete\n",
            "27.1 % complete\n",
            "27.200000000000003 % complete\n",
            "27.3 % complete\n",
            "27.400000000000002 % complete\n",
            "27.500000000000004 % complete\n",
            "27.6 % complete\n",
            "27.700000000000003 % complete\n",
            "27.800000000000004 % complete\n",
            "27.900000000000002 % complete\n",
            "28.000000000000004 % complete\n",
            "28.1 % complete\n",
            "28.199999999999996 % complete\n",
            "28.299999999999997 % complete\n",
            "28.4 % complete\n",
            "28.499999999999996 % complete\n",
            "28.599999999999998 % complete\n",
            "28.7 % complete\n",
            "28.799999999999997 % complete\n",
            "28.9 % complete\n",
            "28.999999999999996 % complete\n",
            "29.099999999999998 % complete\n",
            "29.2 % complete\n",
            "29.299999999999997 % complete\n",
            "29.4 % complete\n",
            "29.5 % complete\n",
            "29.599999999999998 % complete\n",
            "29.7 % complete\n",
            "29.799999999999997 % complete\n",
            "29.9 % complete\n",
            "30.0 % complete\n",
            "30.099999999999998 % complete\n",
            "30.2 % complete\n",
            "30.3 % complete\n",
            "30.4 % complete\n",
            "30.5 % complete\n",
            "30.599999999999998 % complete\n",
            "30.7 % complete\n",
            "30.8 % complete\n",
            "30.9 % complete\n",
            "31.0 % complete\n",
            "31.1 % complete\n",
            "31.2 % complete\n",
            "31.3 % complete\n",
            "31.4 % complete\n",
            "31.5 % complete\n",
            "31.6 % complete\n",
            "31.7 % complete\n",
            "31.8 % complete\n",
            "31.900000000000002 % complete\n",
            "32.0 % complete\n",
            "32.1 % complete\n",
            "32.2 % complete\n",
            "32.300000000000004 % complete\n",
            "32.4 % complete\n",
            "32.5 % complete\n",
            "32.6 % complete\n",
            "32.7 % complete\n",
            "32.800000000000004 % complete\n",
            "32.9 % complete\n",
            "33.0 % complete\n",
            "33.1 % complete\n",
            "33.2 % complete\n",
            "33.300000000000004 % complete\n",
            "33.4 % complete\n",
            "33.5 % complete\n",
            "33.6 % complete\n",
            "33.7 % complete\n",
            "33.800000000000004 % complete\n",
            "33.900000000000006 % complete\n",
            "34.0 % complete\n",
            "34.1 % complete\n",
            "34.2 % complete\n",
            "34.300000000000004 % complete\n",
            "34.4 % complete\n",
            "34.5 % complete\n",
            "34.599999999999994 % complete\n",
            "34.699999999999996 % complete\n",
            "34.8 % complete\n",
            "34.9 % complete\n",
            "35.0 % complete\n",
            "35.099999999999994 % complete\n",
            "35.199999999999996 % complete\n",
            "35.3 % complete\n",
            "35.4 % complete\n",
            "35.5 % complete\n",
            "35.6 % complete\n",
            "35.699999999999996 % complete\n",
            "35.8 % complete\n",
            "35.9 % complete\n",
            "36.0 % complete\n",
            "36.1 % complete\n",
            "36.199999999999996 % complete\n",
            "36.3 % complete\n",
            "36.4 % complete\n",
            "36.5 % complete\n",
            "36.6 % complete\n",
            "36.7 % complete\n",
            "36.8 % complete\n",
            "36.9 % complete\n",
            "37.0 % complete\n",
            "37.1 % complete\n",
            "37.2 % complete\n",
            "37.3 % complete\n",
            "37.4 % complete\n",
            "37.5 % complete\n",
            "37.6 % complete\n",
            "37.7 % complete\n",
            "37.8 % complete\n",
            "37.9 % complete\n",
            "38.0 % complete\n",
            "38.1 % complete\n",
            "38.2 % complete\n",
            "38.3 % complete\n",
            "38.4 % complete\n",
            "38.5 % complete\n",
            "38.6 % complete\n",
            "38.7 % complete\n",
            "38.800000000000004 % complete\n",
            "38.9 % complete\n",
            "39.0 % complete\n",
            "39.1 % complete\n",
            "39.2 % complete\n",
            "39.300000000000004 % complete\n",
            "39.4 % complete\n",
            "39.5 % complete\n",
            "39.6 % complete\n",
            "39.7 % complete\n",
            "39.800000000000004 % complete\n",
            "39.900000000000006 % complete\n",
            "40.0 % complete\n",
            "40.1 % complete\n",
            "40.2 % complete\n",
            "40.300000000000004 % complete\n",
            "40.400000000000006 % complete\n",
            "40.5 % complete\n",
            "40.6 % complete\n",
            "40.699999999999996 % complete\n",
            "40.8 % complete\n",
            "40.9 % complete\n",
            "41.0 % complete\n",
            "41.099999999999994 % complete\n",
            "41.199999999999996 % complete\n",
            "41.3 % complete\n",
            "41.4 % complete\n",
            "41.5 % complete\n",
            "41.6 % complete\n",
            "41.699999999999996 % complete\n",
            "41.8 % complete\n",
            "41.9 % complete\n",
            "42.0 % complete\n",
            "42.1 % complete\n",
            "42.199999999999996 % complete\n",
            "42.3 % complete\n",
            "42.4 % complete\n",
            "42.5 % complete\n",
            "42.6 % complete\n",
            "42.699999999999996 % complete\n",
            "42.8 % complete\n",
            "42.9 % complete\n",
            "43.0 % complete\n",
            "43.1 % complete\n",
            "43.2 % complete\n",
            "43.3 % complete\n",
            "43.4 % complete\n",
            "43.5 % complete\n",
            "43.6 % complete\n",
            "43.7 % complete\n",
            "43.8 % complete\n",
            "43.9 % complete\n",
            "44.0 % complete\n",
            "44.1 % complete\n",
            "44.2 % complete\n",
            "44.3 % complete\n",
            "44.4 % complete\n",
            "44.5 % complete\n",
            "44.6 % complete\n",
            "44.7 % complete\n",
            "44.800000000000004 % complete\n",
            "44.9 % complete\n",
            "45.0 % complete\n",
            "45.1 % complete\n",
            "45.2 % complete\n",
            "45.300000000000004 % complete\n",
            "45.4 % complete\n",
            "45.5 % complete\n",
            "45.6 % complete\n",
            "45.7 % complete\n",
            "45.800000000000004 % complete\n",
            "45.9 % complete\n",
            "46.0 % complete\n",
            "46.1 % complete\n",
            "46.2 % complete\n",
            "46.300000000000004 % complete\n",
            "46.400000000000006 % complete\n",
            "46.5 % complete\n",
            "46.6 % complete\n",
            "46.7 % complete\n",
            "46.800000000000004 % complete\n",
            "46.9 % complete\n",
            "47.0 % complete\n",
            "47.099999999999994 % complete\n",
            "47.199999999999996 % complete\n",
            "47.3 % complete\n",
            "47.4 % complete\n",
            "47.5 % complete\n",
            "47.599999999999994 % complete\n",
            "47.699999999999996 % complete\n",
            "47.8 % complete\n",
            "47.9 % complete\n",
            "48.0 % complete\n",
            "48.1 % complete\n",
            "48.199999999999996 % complete\n",
            "48.3 % complete\n",
            "48.4 % complete\n",
            "48.5 % complete\n",
            "48.6 % complete\n",
            "48.699999999999996 % complete\n",
            "48.8 % complete\n",
            "48.9 % complete\n",
            "49.0 % complete\n",
            "49.1 % complete\n",
            "49.2 % complete\n",
            "49.3 % complete\n",
            "49.4 % complete\n",
            "49.5 % complete\n",
            "49.6 % complete\n",
            "49.7 % complete\n",
            "49.8 % complete\n",
            "49.9 % complete\n",
            "50.0 % complete\n",
            "50.1 % complete\n",
            "50.2 % complete\n",
            "50.3 % complete\n",
            "50.4 % complete\n",
            "50.5 % complete\n",
            "50.6 % complete\n",
            "50.7 % complete\n",
            "50.8 % complete\n",
            "50.9 % complete\n",
            "51.0 % complete\n",
            "51.1 % complete\n",
            "51.2 % complete\n",
            "51.300000000000004 % complete\n",
            "51.4 % complete\n",
            "51.5 % complete\n",
            "51.6 % complete\n",
            "51.7 % complete\n",
            "51.800000000000004 % complete\n",
            "51.9 % complete\n",
            "52.0 % complete\n",
            "52.1 % complete\n",
            "52.2 % complete\n",
            "52.300000000000004 % complete\n",
            "52.400000000000006 % complete\n",
            "52.5 % complete\n",
            "52.6 % complete\n",
            "52.7 % complete\n",
            "52.800000000000004 % complete\n",
            "52.900000000000006 % complete\n",
            "53.0 % complete\n",
            "53.1 % complete\n",
            "53.2 % complete\n",
            "53.300000000000004 % complete\n",
            "53.400000000000006 % complete\n",
            "53.5 % complete\n",
            "53.6 % complete\n",
            "53.7 % complete\n",
            "53.800000000000004 % complete\n",
            "53.900000000000006 % complete\n",
            "54.0 % complete\n",
            "54.1 % complete\n",
            "54.2 % complete\n",
            "54.300000000000004 % complete\n",
            "54.400000000000006 % complete\n",
            "54.50000000000001 % complete\n",
            "54.6 % complete\n",
            "54.7 % complete\n",
            "54.800000000000004 % complete\n",
            "54.900000000000006 % complete\n",
            "55.00000000000001 % complete\n",
            "55.1 % complete\n",
            "55.2 % complete\n",
            "55.300000000000004 % complete\n",
            "55.400000000000006 % complete\n",
            "55.50000000000001 % complete\n",
            "55.60000000000001 % complete\n",
            "55.7 % complete\n",
            "55.800000000000004 % complete\n",
            "55.900000000000006 % complete\n",
            "56.00000000000001 % complete\n",
            "56.10000000000001 % complete\n",
            "56.2 % complete\n",
            "56.3 % complete\n",
            "56.39999999999999 % complete\n",
            "56.49999999999999 % complete\n",
            "56.599999999999994 % complete\n",
            "56.699999999999996 % complete\n",
            "56.8 % complete\n",
            "56.89999999999999 % complete\n",
            "56.99999999999999 % complete\n",
            "57.099999999999994 % complete\n",
            "57.199999999999996 % complete\n",
            "57.3 % complete\n",
            "57.4 % complete\n",
            "57.49999999999999 % complete\n",
            "57.599999999999994 % complete\n",
            "57.699999999999996 % complete\n",
            "57.8 % complete\n",
            "57.9 % complete\n",
            "57.99999999999999 % complete\n",
            "58.099999999999994 % complete\n",
            "58.199999999999996 % complete\n",
            "58.3 % complete\n",
            "58.4 % complete\n",
            "58.5 % complete\n",
            "58.599999999999994 % complete\n",
            "58.699999999999996 % complete\n",
            "58.8 % complete\n",
            "58.9 % complete\n",
            "59.0 % complete\n",
            "59.099999999999994 % complete\n",
            "59.199999999999996 % complete\n",
            "59.3 % complete\n",
            "59.4 % complete\n",
            "59.5 % complete\n",
            "59.599999999999994 % complete\n",
            "59.699999999999996 % complete\n",
            "59.8 % complete\n",
            "59.9 % complete\n",
            "60.0 % complete\n",
            "60.099999999999994 % complete\n",
            "60.199999999999996 % complete\n",
            "60.3 % complete\n",
            "60.4 % complete\n",
            "60.5 % complete\n",
            "60.6 % complete\n",
            "60.699999999999996 % complete\n",
            "60.8 % complete\n",
            "60.9 % complete\n",
            "61.0 % complete\n",
            "61.1 % complete\n",
            "61.199999999999996 % complete\n",
            "61.3 % complete\n",
            "61.4 % complete\n",
            "61.5 % complete\n",
            "61.6 % complete\n",
            "61.7 % complete\n",
            "61.8 % complete\n",
            "61.9 % complete\n",
            "62.0 % complete\n",
            "62.1 % complete\n",
            "62.2 % complete\n",
            "62.3 % complete\n",
            "62.4 % complete\n",
            "62.5 % complete\n",
            "62.6 % complete\n",
            "62.7 % complete\n",
            "62.8 % complete\n",
            "62.9 % complete\n",
            "63.0 % complete\n",
            "63.1 % complete\n",
            "63.2 % complete\n",
            "63.3 % complete\n",
            "63.4 % complete\n",
            "63.5 % complete\n",
            "63.6 % complete\n",
            "63.7 % complete\n",
            "63.800000000000004 % complete\n",
            "63.9 % complete\n",
            "64.0 % complete\n",
            "64.1 % complete\n",
            "64.2 % complete\n",
            "64.3 % complete\n",
            "64.4 % complete\n",
            "64.5 % complete\n",
            "64.60000000000001 % complete\n",
            "64.7 % complete\n",
            "64.8 % complete\n",
            "64.9 % complete\n",
            "65.0 % complete\n",
            "65.10000000000001 % complete\n",
            "65.2 % complete\n",
            "65.3 % complete\n",
            "65.4 % complete\n",
            "65.5 % complete\n",
            "65.60000000000001 % complete\n",
            "65.7 % complete\n",
            "65.8 % complete\n",
            "65.9 % complete\n",
            "66.0 % complete\n",
            "66.10000000000001 % complete\n",
            "66.2 % complete\n",
            "66.3 % complete\n",
            "66.4 % complete\n",
            "66.5 % complete\n",
            "66.60000000000001 % complete\n",
            "66.7 % complete\n",
            "66.8 % complete\n",
            "66.9 % complete\n",
            "67.0 % complete\n",
            "67.10000000000001 % complete\n",
            "67.2 % complete\n",
            "67.30000000000001 % complete\n",
            "67.4 % complete\n",
            "67.5 % complete\n",
            "67.60000000000001 % complete\n",
            "67.7 % complete\n",
            "67.80000000000001 % complete\n",
            "67.9 % complete\n",
            "68.0 % complete\n",
            "68.10000000000001 % complete\n",
            "68.2 % complete\n",
            "68.30000000000001 % complete\n",
            "68.4 % complete\n",
            "68.5 % complete\n",
            "68.60000000000001 % complete\n",
            "68.7 % complete\n",
            "68.8 % complete\n",
            "68.89999999999999 % complete\n",
            "69.0 % complete\n",
            "69.1 % complete\n",
            "69.19999999999999 % complete\n",
            "69.3 % complete\n",
            "69.39999999999999 % complete\n",
            "69.5 % complete\n",
            "69.6 % complete\n",
            "69.69999999999999 % complete\n",
            "69.8 % complete\n",
            "69.89999999999999 % complete\n",
            "70.0 % complete\n",
            "70.1 % complete\n",
            "70.19999999999999 % complete\n",
            "70.3 % complete\n",
            "70.39999999999999 % complete\n",
            "70.5 % complete\n",
            "70.6 % complete\n",
            "70.7 % complete\n",
            "70.8 % complete\n",
            "70.89999999999999 % complete\n",
            "71.0 % complete\n",
            "71.1 % complete\n",
            "71.2 % complete\n",
            "71.3 % complete\n",
            "71.39999999999999 % complete\n",
            "71.5 % complete\n",
            "71.6 % complete\n",
            "71.7 % complete\n",
            "71.8 % complete\n",
            "71.89999999999999 % complete\n",
            "72.0 % complete\n",
            "72.1 % complete\n",
            "72.2 % complete\n",
            "72.3 % complete\n",
            "72.39999999999999 % complete\n",
            "72.5 % complete\n",
            "72.6 % complete\n",
            "72.7 % complete\n",
            "72.8 % complete\n",
            "72.89999999999999 % complete\n",
            "73.0 % complete\n",
            "73.1 % complete\n",
            "73.2 % complete\n",
            "73.3 % complete\n",
            "73.4 % complete\n",
            "73.5 % complete\n",
            "73.6 % complete\n",
            "73.7 % complete\n",
            "73.8 % complete\n",
            "73.9 % complete\n",
            "74.0 % complete\n",
            "74.1 % complete\n",
            "74.2 % complete\n",
            "74.3 % complete\n",
            "74.4 % complete\n",
            "74.5 % complete\n",
            "74.6 % complete\n",
            "74.7 % complete\n",
            "74.8 % complete\n",
            "74.9 % complete\n",
            "75.0 % complete\n",
            "75.1 % complete\n",
            "75.2 % complete\n",
            "75.3 % complete\n",
            "75.4 % complete\n",
            "75.5 % complete\n",
            "75.6 % complete\n",
            "75.7 % complete\n",
            "75.8 % complete\n",
            "75.9 % complete\n",
            "76.0 % complete\n",
            "76.1 % complete\n",
            "76.2 % complete\n",
            "76.3 % complete\n",
            "76.4 % complete\n",
            "76.5 % complete\n",
            "76.6 % complete\n",
            "76.7 % complete\n",
            "76.8 % complete\n",
            "76.9 % complete\n",
            "77.0 % complete\n",
            "77.10000000000001 % complete\n",
            "77.2 % complete\n",
            "77.3 % complete\n",
            "77.4 % complete\n",
            "77.5 % complete\n",
            "77.60000000000001 % complete\n",
            "77.7 % complete\n",
            "77.8 % complete\n",
            "77.9 % complete\n",
            "78.0 % complete\n",
            "78.10000000000001 % complete\n",
            "78.2 % complete\n",
            "78.3 % complete\n",
            "78.4 % complete\n",
            "78.5 % complete\n",
            "78.60000000000001 % complete\n",
            "78.7 % complete\n",
            "78.8 % complete\n",
            "78.9 % complete\n",
            "79.0 % complete\n",
            "79.10000000000001 % complete\n",
            "79.2 % complete\n",
            "79.3 % complete\n",
            "79.4 % complete\n",
            "79.5 % complete\n",
            "79.60000000000001 % complete\n",
            "79.7 % complete\n",
            "79.80000000000001 % complete\n",
            "79.9 % complete\n",
            "80.0 % complete\n",
            "80.10000000000001 % complete\n",
            "80.2 % complete\n",
            "80.30000000000001 % complete\n",
            "80.4 % complete\n",
            "80.5 % complete\n",
            "80.60000000000001 % complete\n",
            "80.7 % complete\n",
            "80.80000000000001 % complete\n",
            "80.9 % complete\n",
            "81.0 % complete\n",
            "81.10000000000001 % complete\n",
            "81.2 % complete\n",
            "81.3 % complete\n",
            "81.39999999999999 % complete\n",
            "81.5 % complete\n",
            "81.6 % complete\n",
            "81.69999999999999 % complete\n",
            "81.8 % complete\n",
            "81.89999999999999 % complete\n",
            "82.0 % complete\n",
            "82.1 % complete\n",
            "82.19999999999999 % complete\n",
            "82.3 % complete\n",
            "82.39999999999999 % complete\n",
            "82.5 % complete\n",
            "82.6 % complete\n",
            "82.69999999999999 % complete\n",
            "82.8 % complete\n",
            "82.89999999999999 % complete\n",
            "83.0 % complete\n",
            "83.1 % complete\n",
            "83.2 % complete\n",
            "83.3 % complete\n",
            "83.39999999999999 % complete\n",
            "83.5 % complete\n",
            "83.6 % complete\n",
            "83.7 % complete\n",
            "83.8 % complete\n",
            "83.89999999999999 % complete\n",
            "84.0 % complete\n",
            "84.1 % complete\n",
            "84.2 % complete\n",
            "84.3 % complete\n",
            "84.39999999999999 % complete\n",
            "84.5 % complete\n",
            "84.6 % complete\n",
            "84.7 % complete\n",
            "84.8 % complete\n",
            "84.89999999999999 % complete\n",
            "85.0 % complete\n",
            "85.1 % complete\n",
            "85.2 % complete\n",
            "85.3 % complete\n",
            "85.39999999999999 % complete\n",
            "85.5 % complete\n",
            "85.6 % complete\n",
            "85.7 % complete\n",
            "85.8 % complete\n",
            "85.9 % complete\n",
            "86.0 % complete\n",
            "86.1 % complete\n",
            "86.2 % complete\n",
            "86.3 % complete\n",
            "86.4 % complete\n",
            "86.5 % complete\n",
            "86.6 % complete\n",
            "86.7 % complete\n",
            "86.8 % complete\n",
            "86.9 % complete\n",
            "87.0 % complete\n",
            "87.1 % complete\n",
            "87.2 % complete\n",
            "87.3 % complete\n",
            "87.4 % complete\n",
            "87.5 % complete\n",
            "87.6 % complete\n",
            "87.7 % complete\n",
            "87.8 % complete\n",
            "87.9 % complete\n",
            "88.0 % complete\n",
            "88.1 % complete\n",
            "88.2 % complete\n",
            "88.3 % complete\n",
            "88.4 % complete\n",
            "88.5 % complete\n",
            "88.6 % complete\n",
            "88.7 % complete\n",
            "88.8 % complete\n",
            "88.9 % complete\n",
            "89.0 % complete\n",
            "89.1 % complete\n",
            "89.2 % complete\n",
            "89.3 % complete\n",
            "89.4 % complete\n",
            "89.5 % complete\n",
            "89.60000000000001 % complete\n",
            "89.7 % complete\n",
            "89.8 % complete\n",
            "89.9 % complete\n",
            "90.0 % complete\n",
            "90.10000000000001 % complete\n",
            "90.2 % complete\n",
            "90.3 % complete\n",
            "90.4 % complete\n",
            "90.5 % complete\n",
            "90.60000000000001 % complete\n",
            "90.7 % complete\n",
            "90.8 % complete\n",
            "90.9 % complete\n",
            "91.0 % complete\n",
            "91.10000000000001 % complete\n",
            "91.2 % complete\n",
            "91.3 % complete\n",
            "91.4 % complete\n",
            "91.5 % complete\n",
            "91.60000000000001 % complete\n",
            "91.7 % complete\n",
            "91.8 % complete\n",
            "91.9 % complete\n",
            "92.0 % complete\n",
            "92.10000000000001 % complete\n",
            "92.2 % complete\n",
            "92.30000000000001 % complete\n",
            "92.4 % complete\n",
            "92.5 % complete\n",
            "92.60000000000001 % complete\n",
            "92.7 % complete\n",
            "92.80000000000001 % complete\n",
            "92.9 % complete\n",
            "93.0 % complete\n",
            "93.10000000000001 % complete\n",
            "93.2 % complete\n",
            "93.30000000000001 % complete\n",
            "93.4 % complete\n",
            "93.5 % complete\n",
            "93.60000000000001 % complete\n",
            "93.7 % complete\n",
            "93.8 % complete\n",
            "93.89999999999999 % complete\n",
            "94.0 % complete\n",
            "94.1 % complete\n",
            "94.19999999999999 % complete\n",
            "94.3 % complete\n",
            "94.39999999999999 % complete\n",
            "94.5 % complete\n",
            "94.6 % complete\n",
            "94.69999999999999 % complete\n",
            "94.8 % complete\n",
            "94.89999999999999 % complete\n",
            "95.0 % complete\n",
            "95.1 % complete\n",
            "95.19999999999999 % complete\n",
            "95.3 % complete\n",
            "95.39999999999999 % complete\n",
            "95.5 % complete\n",
            "95.6 % complete\n",
            "95.7 % complete\n",
            "95.8 % complete\n",
            "95.89999999999999 % complete\n",
            "96.0 % complete\n",
            "96.1 % complete\n",
            "96.2 % complete\n",
            "96.3 % complete\n",
            "96.39999999999999 % complete\n",
            "96.5 % complete\n",
            "96.6 % complete\n",
            "96.7 % complete\n",
            "96.8 % complete\n",
            "96.89999999999999 % complete\n",
            "97.0 % complete\n",
            "97.1 % complete\n",
            "97.2 % complete\n",
            "97.3 % complete\n",
            "97.39999999999999 % complete\n",
            "97.5 % complete\n",
            "97.6 % complete\n",
            "97.7 % complete\n",
            "97.8 % complete\n",
            "97.89999999999999 % complete\n",
            "98.0 % complete\n",
            "98.1 % complete\n",
            "98.2 % complete\n",
            "98.3 % complete\n",
            "98.4 % complete\n",
            "98.5 % complete\n",
            "98.6 % complete\n",
            "98.7 % complete\n",
            "98.8 % complete\n",
            "98.9 % complete\n",
            "99.0 % complete\n",
            "99.1 % complete\n",
            "99.2 % complete\n",
            "99.3 % complete\n",
            "99.4 % complete\n",
            "99.5 % complete\n",
            "99.6 % complete\n",
            "99.7 % complete\n",
            "99.8 % complete\n",
            "99.9 % complete\n",
            "100.0 % complete\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Example ndarray conversion:\n",
        "\n",
        "a = np.array([[0, 1, 0, 0], [1, 0, 0, 0]])\n",
        "print(a)\n",
        "print(type(a))\n",
        "print(a.shape)\n",
        "b = a.tolist()\n",
        "s = json.dumps(b)\n",
        "arr = np.array(json.loads(s))\n",
        "print(arr)\n",
        "print(type(arr))\n",
        "print(arr.shape)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2yheHe7tFLuf",
        "outputId": "d9db7ce0-2ded-4513-b564-91d154712a10"
      },
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[[0 1 0 0]\n",
            " [1 0 0 0]]\n",
            "<class 'numpy.ndarray'>\n",
            "(2, 4)\n",
            "[[0 1 0 0]\n",
            " [1 0 0 0]]\n",
            "<class 'numpy.ndarray'>\n",
            "(2, 4)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Final check that df is well-formed\n",
        "df\n",
        "print(len(df.iloc[0]['Seq']))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "RRppzokLaBmV",
        "outputId": "cda6ca78-a379-4286-d4cd-2c046c9206d8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "3822\n"
          ]
        }
      ]
    }
  ]
}